### PR TITLE
Steinhardt class: A new single class for the Steinhardt order Parameter

### PR DESCRIPTION
**Pull request by b-butler**
## Description
The goal is to create a class Steinhardt that better emulates the functionality
of other classes within freud.order with the Steinhardt order parameter. This
functionality is currently encapsulated into the LocalQl and LocalWl classes.
This new class would in theory become the new standard API will leaving the
other classes for backwards-compatibility.

Some goals for this branch are ...
+ Provide a single compute method in the python interface
+ Use flags at construction to determine the particular order parameter given by
  compute.
+ Keep current and improve further - when applicable - performance of
  class with an eye towards potenial further parallelization.
+ Improve cpp code readability and better emulate the DRY mantra.

## Motivation and Context
Most other classes within the order module have a single compute method exposed
to the python interface. However, with the Steinhardt order parameters currently
8 exist between 2 classes. Furthermore, the C++ code use to determine the
Steinhardt order parameter requires code repetition. The current python
interface also exposes the user to this complexity. Until recently, the logic
for the compute methods were confusingly split between the Cython and C++
layers. Further improvements to code readability and maintainability are
warrented and this new API would allow for that.

## Pull Request Status
- [x] Create draft Steinhardt.cc. This should initially mostly be copying code
  from LocalQl and LocalWl to file.
- [x] Create working compute method.
  - [x] Impliment logic using member flags to calculate only what is necessary.
  - [x] Abstract array reallocation to function that only reallocates used
    arrays.
  - [x] Potentially move individual computational steps in compute method to
    private functions.
  - [ ] Add parallelization to loops over particle number.
- [x] Clean up code where applicable.
- [x] Add Steinhardt class to \_order.pxd.
- [x] Add Steinhardt class to order.pyx.
- [x] Create tests for Steinhardt class.
- [x] Update documentation

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc:1]: Can you retab this file? It looks like in a lot of places you end up with 8 spaces instead of 4 for tabs.

Comment (vyasr) [reply cpp/order/Steinhardt.cc:1]: Also make sure you follow our (admittedly strange) code style with respect to brace indentation, see [here](https://freud.readthedocs.io/en/latest/sourceconventions.html#cpp)

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: What check isn't necessary? Is there something we can remove here, or is this a copied comment from old code?

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Remove

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Is this your comment?

File: cpp/order/Steinhardt.h
Comment (vyasr) [inline cpp/order/Steinhardt.h]: Fix indentation

File: cpp/order/Steinhardt.h
Comment (vyasr) [inline cpp/order/Steinhardt.h]: Unindent `private`

Comment (vyasr) [reply cpp/order/Steinhardt.h]: Also remove a couple of the extra blank lines above

File: cpp/order/Steinhardt.h
Comment (vyasr) [inline cpp/order/Steinhardt.h]: Add a space `// if any`

File: freud/order.pyx
Comment (vyasr) [inline freud/order.pyx]: type `argument`

File: freud/order.pyx
Comment (vyasr) [inline freud/order.pyx:533]: Add yourself! :)

File: freud/order.pyx
Comment (vyasr) [inline freud/order.pyx]: typo `calculate`

File: tests/test_order_Steinhardt.py
Comment (vyasr) [inline tests/test_order_Steinhardt.py]: For consistency with our other tests, I would remove this here and in other tests whether it's still present.

**Review by vyasr — CHANGES_REQUESTED** [review]
There are some style issues, but the content all looks fine to me. There are a few places where it would be nice to combine a little more of the logic, but unfortunately different code paths differ just enough that it's not possible.

File: cpp/order/Steinhardt.h
Comment (bdice) [inline cpp/order/Steinhardt.h]: typo `necessary`

File: freud/_order.pxd
Comment (bdice) [inline freud/_order.pxd]: You'll need to fix the code here to address the `flake8` code style issues that are preventing tests from running. See here for info about the `flake8` pre-commit hook, which can help you catch these early: https://freud.readthedocs.io/en/stable/sourceconventions.html#python

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx]: `W_l` needs an underscore

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx]: Fix the incomplete sentence here. Maybe write "Normalization is performed by..."

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx]: We'll want to actually make this change - the `box` shouldn't be in the constructor anymore, only in the `compute` method.

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx]: I discussed with @vyasr a while back - I think we'd like to use the property name `order` for all order parameters' data going forward. Since this is a new class, let's rename `St` to `order`.

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx:604]: I'd like to see this logic implemented on the C++ level.

File: freud/order.pyx
Comment (bdice) [inline freud/order.pyx]: We'll probably refactor this to use the new neighbor-finding code after #272 has been merged. Then we will handle the neighbor-finding logic (e.g. `rmax`, `rmin`, `k` nearest neighbors) via a `QueryArgs` argument to the `compute` method. That way, we'll also encompass the behavior of the current `LocalQlNear` and `LocalWlNear` classes.

**Comment by bdice** [comment]
@b-butler Thanks for your work on this! It will be a really nice feature improvement. We'll do a few iterations on this to get it right, since this feature is so heavily used. Ping @vyasr and I if you have questions or when you've made a round of changes and are ready for more review.

File: freud/order.pyx
Comment (vyasr) [reply freud/order.pyx:604]: To clarify, you want there to be a single function on the C++ side that returns the correct order parameter depending on whether or not the flag indicates to use Wl or Ql? Keep in mind that the two currently return different types of output, so I'm not sure you can accomplish that.

I also don't think the Wl really needs to be complex from what I recall of the math defining it, but IIRC you do get meaningful imaginary components from the Wl classes. I haven't really used them in years, so I don't know what to make of that, and I would suggest holding off making any changes to the types, etc, for a separate PR where we do more explicit validation checking against e.g. pyboo.

File: freud/order.pyx
Comment (vyasr) [reply freud/order.pyx]: Agreed. For now it's fine that this only supports the distance cutoffs.

File: cpp/order/Steinhardt.cc
Comment (b-butler) [reply cpp/order/Steinhardt.cc:1]: Will do

File: cpp/order/Steinhardt.cc
Comment (b-butler) [reply cpp/order/Steinhardt.cc]: Copied from old code

File: cpp/order/Steinhardt.cc
Comment (b-butler) [reply cpp/order/Steinhardt.cc]: No that I think was left over from the hackathon, but I must have missed it

File: freud/order.pyx
Comment (b-butler) [reply freud/order.pyx:604]: I would have liked to impliment the logic in C++, but vyasr is right. The type difference between Wl and Ql prevents the logic from being on the C++ side.

**Comment by b-butler** [comment]
I addressed all the comments and fixed some other stuff I noticed.

**Comment by bdice** [comment]
@b-butler Good work so far! A few broad tasks for refactoring and simplification:

1. Many of the computations (e.g. normalization) are repeated through the code. For example, I would prefer to see one "normalize" call that accepts shared pointers as arguments and uses the flags of the class to determine how to specialize, instead of `computeNorm`, `computeAveNorm`, `computeNormWl`, `computeAveNormWl`.
2. If the `norm` flag is false, do we need to compute `m_Qlm` at all? From what I can tell, this data is only used if `norm` is true. It would be good to verify that and make that step conditional if possible.
3. There is a loop in `computeAve` that is labeled "Can be parallelized." It would be good to parallelize this.

File: freud/order.pyx
Comment (bdice) [reply freud/order.pyx:604]: Just saw this - editing my last comment about next steps since this isn't possible. Sorry for the confusion.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: It doesn't appear like this result is being used - I looked at the original `LocalWl` to confirm. The original code had an option for `enableNormalization` that causes this to be useful, but it doesn't appear that it was ever possible to control this on the Python side. I will remove this in the current changes I am making to the code.

**Comment by bdice** [comment]
I did some clean up and reduced the code repetition for the `Wl` methods. There is still more to be done with reducing repetition, but it may not be exactly what I was originally saying above about `norm` methods. Suggestions 2 and 3 are still valid. Here's the task list I have in mind:
- [ ] Only compute Qlm if needed
- [ ] Parallelize `computeAve` loop
- [ ] Refactor remaining blocks of repeated code into functions
- [ ] Remove unnecessary member variable arrays (e.g. it doesn't make sense to have multiple member variables for `m_WliAveNorm` ... `m_Wli`, etc. if only one of them can be used at a time).

I'm thinking about creating a class that wraps around our shared pointer arrays and plays nicely with Cython, taking the `makeArray` method I just wrote for this PR to the next level. We might combine some nice behaviors of shared pointers and `std::vector` to create a wrapper layer that handles things like using `memset` to zero out the data, easily instantiating an array with a default deleter, making it simpler to iterate over the data, etc. This kind of code is repeated lots of places in freud and it's a prime target for refactoring in my opinion just because it would help simplify our codebase a lot.  Thoughts, @vyasr @jinsoo960?

**Comment by bdice** [comment]
It's strange to me that we return full arrays for the "normalized" order parameters - it's exactly the same value for every particle.

File: cpp/order/Steinhardt.cc
Comment (bdice) [reply cpp/order/Steinhardt.cc]: ^ done.

**Comment by bdice** [comment]
I finished the refactoring tasks I wanted to do - @b-butler, please take a look and feel free to pick up with the remaining tasks when you have time.

**Comment by b-butler** [comment]
Will do @bdice. Just saw all your commits and recommendations.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix spacing here.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: I'd delete this comment, it's referring to a pre-fsph implementation of Steinhardt that must be 3 years old at this point.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix spacing.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: You shouldn't need this qualification, just `reallocateArrays` is enough.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Same as above.

Comment (vyasr) [reply cpp/order/Steinhardt.cc]: I see this in a number of places, I'd just drop all unnecessary `Steinhardt::`.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Shouldn't this already have been zeroed by `reallocateArrays`?

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix spacing.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix spacing.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: I'm confused as to why this conditional is here.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix spacing.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: These should be set by `reallocateArrays`.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix the indentation of this block, and I think you can also get rid of the memset since `reallocateArrays` does it.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Fix indentation.

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Indentation

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Dealt with in reallocateArrays

File: cpp/order/Steinhardt.cc
Comment (vyasr) [inline cpp/order/Steinhardt.cc]: Remove in favor of reallocateArrays

File: cpp/order/Steinhardt.h
Comment (vyasr) [inline cpp/order/Steinhardt.h:9]: I think you can remove some of these (cstring, stdexcept).

**Review by vyasr — CHANGES_REQUESTED** [review]
This is nice, much shorter than the older methods and will significantly clean up that whole chunk of code. I still have some change requests, but it's getting there!

The formatting is still all over the place. Since you first started working on this PR, we've started using clang-format as our C++ style requirement. Can you merge in master and then try to apply that clang-format to your code (let Bradley or me know if you need help doing that)? That should help fix a lot of the spacing issues.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: Seems like this `if` is redundant?

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: See if tests still pass if you remove this `if` statement. We have a problem with this check, see #297. We may need to change the preprocessing we do to the NeighborList to correct the behavior after removing this line.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: Fix comment, `from i pointing to neighbor n`.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: See above comment about removing this check.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc:138]: Add comment: `// Initialize thread-local m_Qlm and compute it in this function, if we won't average over neighbors later`

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: Suggesting a change from `QlmAve_exists` to `Qlm_exists` as above, so it's clear these represent the same thing.

File: cpp/order/Steinhardt.cc
Comment (bdice) [inline cpp/order/Steinhardt.cc]: Move this below the `if` statement (so it's only specified once).

File: cpp/order/Steinhardt.cc
Comment (bdice) [reply cpp/order/Steinhardt.cc:138]: Add a similar comment below in the average method, something like `// Initialize thread-local m_Qlm for storing neighbor-averaged QlmiAve, reduced over particles`

File: cpp/order/Steinhardt.cc
Comment (bdice) [reply cpp/order/Steinhardt.cc]: Specifically we need to check the behavior where `rmin > 0`, when there are neighbors within that inner shell distance.

**Review by bdice — CHANGES_REQUESTED** [review]
This is looking great. A few comments left. I'm really happy with how this is turning out, it's much simpler. 🙌

**Comment by b-butler** [comment]
The tests are currently failing in the github tests, but work in my local environment.

**Comment by vyasr** [comment]
@b-butler the failing tests weren't really your fault. You needed to merge up to current master and make a couple minor changes to your calls to `convert_array`, but also there were issues with the `common` module being moved from pure Python to Cython that were causing problems. Should be fixed now. Remind me to review in the next day or so if I haven't yet.

**Comment by b-butler** [comment]
@vyasr thanks for figuring it out!

**Review by vyasr — APPROVED** [review]
@bdice I'm happy with this PR and I think it's ready. If you're happy with this as well, I recommend merging #317 and then making the change here to the enumerable_thread_specific array we use here, then merging in this PR.

**Pull request merged**
